runtime/vm/segment.c

```diff
@@ -411,6 +411,13 @@ static J9MemorySegment * allocateVirtualMemorySegmentInListInternal(J9JavaVM *ja
 
 			segmentList->totalSegmentSize += segment->size;
 
+			if (J9_ARE_ALL_BITS_SET(segment->type, MEMORY_TYPE_CODE)) {
+				/* For CodeCache segments the JIT will later write a TR::CodeCache structure pointer at the begining of the segment.
+				 * Until then, make sure that a potential reader sees a NULL pointer.
+				 */
+				*((UDATA**)segment->heapBase) = NULL;
```

Comment:
I'm wondering if we should issue a write barrier here. This operation is under mutex, but I don't think that guarantees the ordering of writes within the locked section. Assuming this matters at all, this write should probably be done before assigning the heapBase.

Reply:
This case is very similar to this:

https://github.com/eclipse-openj9/openj9/blob/96d48c94cfd03e344af3ac1ad2ca1d4b5ff10c6f/runtime/vm/jnicsup.cpp#L2172-L2177

Reply:
The jniIDs can be read without mutex, which may be what necessitated this pattern. If the segments can only be read under mutex, the issue may not exist here, but I think I'd prefer the safety just in case.

Reply:
Right, and move that up above the assigning of the segment fields. The write barrier will not be expensive (given all the work we're already doing here), so I'd like to see that included as well.

Reply:
I am having problems using `issueWriteBarrier()` (defined in a cpp file) in a `c` file (I was trying to include `AtomicSupport.hpp`)
Is there a similar construct that can be used in c files?

Reply:
Try including omrutilbase.h instead.

Reply:
issueWriteBarrier is the C-callable equivalent of the atomic support inline. Pete's suggestion is correct.